libyul/optimiser/BlockClassFinder.cpp

```diff
@@ -0,0 +1,246 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Optimiser component that combines syntactically equivalent functions.
+ */
+
+#include <libyul/optimiser/BlockClassFinder.h>
+#include <libyul/optimiser/SyntacticalEquality.h>
+#include <libdevcore/CommonData.h>
+
+using namespace std;
+using namespace dev;
+using namespace yul;
+
+namespace {
+static constexpr uint64_t literalHash(char const *_literal, size_t _N) {
+	return (_N == 0) ? 14695981039346656037u : literalHash(_literal + 1, _N - 1) ^ 1099511628211u;
+}
+
+template<size_t N>
+static constexpr uint64_t literalHash(char const (&_literal)[N]) {
+	return literalHash(_literal, N);
+}
+}
+
+bool BlockClassFinder::isExternal(YulString _identifier) const
+{
+	auto it = m_identifierMapping.find(_identifier);
+	yulAssert(it != m_identifierMapping.end(), "");
+	return (it->second & 1) == 0;
+}
+
+std::vector<BlockClass> BlockClassFinder::run(Block const& _block)
+{
+	GlobalState result;
+	BlockClassFinder blockClassFinder(result);
+	blockClassFinder(_block);
+	return std::move(result.blockClasses);
+}
+
+void BlockClassFinder::operator()(Literal const& _literal)
+{
+	hash(literalHash("Literal"));
+	hash(_literal.value.hash());
+	hash(_literal.type.hash());
+	hash(static_cast<std::underlying_type_t<LiteralKind>>(_literal.kind));
+}
+
+void BlockClassFinder::operator()(Identifier const& _identifier)
+{
+	hash(literalHash("Identifier"));
+	size_t id = 0;
+	auto it = m_identifierMapping.find(_identifier.name);
+	if (it == m_identifierMapping.end())
+	{
+		id = 2 * (m_externalIdentifierCount++);
```

Comment:
Why only even numbers?

Reply:
Yeah, currently internal and external references get odd and even IDs to distinguish between them - another thing to be cleaned up :-).